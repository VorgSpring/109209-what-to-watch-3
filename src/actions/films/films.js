import {
  LOAD_FILMS_REQUEST,
  LOAD_FILMS_SUCCESS,
  LOAD_FILMS_ERROR
} from '../../constants/actions-type';

export const loadFilmsRequest = () => ({
  type: LOAD_FILMS_REQUEST
});

export const loadFilmsSuccess = (films) => ({
  type: LOAD_FILMS_SUCCESS,
  payload: films
});

export const loadFilmsError = (error) => ({
  type: LOAD_FILMS_ERROR,
  payload: error
});

