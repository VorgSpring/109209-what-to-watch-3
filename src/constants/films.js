export const MAX_COUNT_SIMILAR_FILMS = 4;

export const ALL_GENRES_TYPE = `All genres`;
