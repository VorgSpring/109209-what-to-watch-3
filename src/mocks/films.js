export const films = [
  {
    id: 1,
    bgImage: `/bg-image`,
    name: `Fantastic Beasts: The Crimes of Grindelwald`,
    poster: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    previewImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    genre: `Crime`,
    released: 2000,
    preview: [
      {
        src: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
        type: `video/mp4`
      },
      {
        src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
        type: `video/webm`
      }
    ],
    isFavorite: false,
    rating: 1,
    director: `blah`,
    scoresCount: 3,
    description: `blah`,
    starring: [
      `blah`, `blah-blah`
    ],
    runTime: 123
  },
  {
    id: 2,
    bgImage: `/bg-image`,
    name: `Bohemian Rhapsody`,
    poster: `img/bohemian-rhapsody.jpg`,
    previewImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    genre: `Romance`,
    released: 2000,
    preview: [
      {
        src: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
        type: `video/mp4`
      },
      {
        src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
        type: `video/webm`
      }
    ],
    isFavorite: false,
    rating: 1,
    director: `blah`,
    scoresCount: 3,
    description: `blah`,
    starring: [
      `blah`, `blah`
    ]
  },
  {
    id: 3,
    bgImage: `/bg-image`,
    name: `Macbeth`,
    poster: `img/macbeth.jpg`,
    previewImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    genre: `Horror`,
    released: 2000,
    preview: [
      {
        src: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
        type: `video/mp4`
      },
      {
        src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
        type: `video/webm`
      }
    ],
    isFavorite: false,
    rating: 1,
    director: `blah`,
    scoresCount: 3,
    description: `blah`,
    starring: [
      `blah`, `blah`
    ]
  },
  {
    id: 4,
    bgImage: `/bg-image`,
    name: `Aviator`,
    poster: `img/aviator.jpg`,
    previewImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    genre: `Dramas`,
    released: 2000,
    preview: [
      {
        src: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
        type: `video/mp4`
      },
      {
        src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
        type: `video/webm`
      }
    ],
    isFavorite: false,
    rating: 1,
    director: `blah`,
    scoresCount: 3,
    description: `blah`,
    starring: [
      `blah`, `blah`
    ]
  },
  {
    id: 5,
    bgImage: `/bg-image`,
    bgColor: `#fff`,
    name: `We need to talk about Kevin`,
    poster: `img/we-need-to-talk-about-kevin.jpg`,
    previewImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    genre: `Documentary`,
    released: 2000,
    preview: [
      {
        src: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
        type: `video/mp4`
      },
      {
        src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
        type: `video/webm`
      }
    ],
    isFavorite: false,
    rating: 1,
    director: `blah`,
    scoresCount: 3,
    description: `blah`,
    starring: [
      `blah`, `blah`
    ],
    runTime: 123
  },
];

/* eslint-disable camelcase */
export const filmByServer = {
  id: 1,
  name: `The Grand Budapest Hotel`,
  poster_image: `img/the-grand-budapest-hotel-poster.jpg`,
  preview_image: `img/the-grand-budapest-hotel.jpg`,
  background_image: `img/the-grand-budapest-hotel-bg.jpg`,
  background_color: `#ffffff`,
  video_link: `https://some-link`,
  preview_video_link: `https://some-link`,
  description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.`,
  rating: 8.9,
  scores_count: 240,
  director: `Wes Andreson`,
  starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Saoirse Ronan`],
  run_time: 99,
  genre: `Comedy`,
  released: 2014,
  is_favorite: false,
};
