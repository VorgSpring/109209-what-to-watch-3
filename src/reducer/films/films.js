import {
  LOAD_FILMS_REQUEST,
  LOAD_FILMS_SUCCESS,
  LOAD_FILMS_ERROR,
  CHANGE_FAVORITES_SUCCESS
} from '../../constants/actions-type';

const initState = {
  isLoading: false,
  isError: false,
  films: null
};

export default (state = initState, action) => {
  switch (action.type) {
    case (LOAD_FILMS_REQUEST):
      return Object.assign(
          {}, state, {
            isLoading: true,
            isError: false
          }
      );

    case (LOAD_FILMS_SUCCESS):
      return Object.assign(
          {}, state, {
            isLoading: false,
            films: action.payload
          }
      );

    case (CHANGE_FAVORITES_SUCCESS):
      return Object.assign(
          {}, state, {
            films: state.films.map((film) => {
              if (film.id === action.payload.id) {
                return action.payload;
              }
              return film;
            })
          }
      );

    case (LOAD_FILMS_ERROR):
      return Object.assign(
          {}, state, {
            isLoading: false,
            isError: true
          }
      );
    default:
      return state;
  }
};
