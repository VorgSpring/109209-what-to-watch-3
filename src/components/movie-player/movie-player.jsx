import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {VideoPlayer} from '../video-player/video-player.jsx';
import {PlayerButton} from '../../constants/player';
import {prepareTimeDataFromSeconds} from '../../helpers/prepared-time-data/prepared-time-data';
import {withVideoPlayer} from '../../hoc/with-video-player/with-video-player';
import {getFilmByIdSelector} from '../../selectors/films/films';

export class MoviePlayerComponent extends PureComponent {
  constructor(props) {
    super(props);

    this.handleClosePlayer = this.handleClosePlayer.bind(this);
  }

  handleClosePlayer() {
    this.props.history.goBack();
  }

  render() {
    const {
      film,
      progress,
      currentTime,
      status,
      toggleVideo,
      setFullscreen,
      refNode
    } = this.props;

    const {
      name,
      previewImage,
      preview
    } = film;

    return (
      <div className='player'>
        <VideoPlayer
          ref={refNode}
          poster={previewImage}
          name={name}
          source={preview}
          autoPlay
        />

        <button
          type='button'
          className='player__exit'
          onClick={this.handleClosePlayer}
        >
          Exit
        </button>

        {status && <div className='player__controls'>
          <div className='player__controls-row'>
            <div className='player__time'>
              <progress
                className='player__progress'
                value={progress}
                max='100' />
              <div
                className='player__toggler'
                style={{
                  left: `${progress}%`
                }}
              >
                Toggler
              </div>
            </div>
            <div className='player__time-value'>
              {prepareTimeDataFromSeconds(currentTime)}
            </div>
          </div>

          <div className='player__controls-row'>
            <button
              type='button'
              className='player__play'
              onClick={toggleVideo}
            >
              <svg viewBox='0 0 19 19' width='19' height='19'>
                <use xlinkHref={PlayerButton[status].ICON} />
              </svg>
              <span>{PlayerButton[status].TEXT}</span>
            </button>
            <div className='player__name'>{name}</div>

            <button
              type='button'
              className='player__full-screen'
              onClick={setFullscreen}>
              <svg viewBox='0 0 27 27' width='27' height='27'>
                <use xlinkHref='#full-screen' />
              </svg>
              <span>Full screen</span>
            </button>
          </div>
        </div>}
      </div>
    );
  }
}

MoviePlayerComponent.defaultProps = {
  status: null
};

MoviePlayerComponent.propTypes = {
  film: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    previewImage: PropTypes.string.isRequired,
    preview: PropTypes.oneOfType([
      PropTypes.array,
      PropTypes.string
    ]).isRequired,
  }).isRequired,
  status: PropTypes.string,
  history: PropTypes.shape({
    goBack: PropTypes.func.isRequired
  }).isRequired,
  refNode: PropTypes.object.isRequired,
  progress: PropTypes.number.isRequired,
  currentTime: PropTypes.number.isRequired,
  toggleVideo: PropTypes.func.isRequired,
  setFullscreen: PropTypes.func.isRequired
};

const MoviePlayerContainer = withVideoPlayer(MoviePlayerComponent);

const mapStateToProps = (state, ownProps) => ({
  film: getFilmByIdSelector(state, ownProps)
});

export const MoviePlayer = connect(
    mapStateToProps
)(MoviePlayerContainer);

