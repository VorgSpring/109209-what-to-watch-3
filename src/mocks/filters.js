import {ALL_GENRES_TYPE} from '../constants/films';

export const filters = {
  genre: ALL_GENRES_TYPE.ALL
};
