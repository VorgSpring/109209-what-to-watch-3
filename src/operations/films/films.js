import {
  loadFilmsRequest,
  loadFilmsSuccess,
  loadFilmsError
} from '../../actions/films/films';
import {ApiPaths} from '../../constants/api';
import {preparedFilmsData} from '../../helpers/prepared-films-data/prepared-films-data';

export const loadFilms = () => (dispatch, _getState, api) => {
  dispatch(loadFilmsRequest());

  return api.get(ApiPaths.FILMS)
    .then((response) => {
      const films = preparedFilmsData(response.data);
      dispatch(loadFilmsSuccess(films));
    })
    .catch((e) => {
      const {error} = e.response.data;
      dispatch(loadFilmsError({error}));
    });
};
