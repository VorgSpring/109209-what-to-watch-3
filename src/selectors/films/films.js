import {createSelector} from 'reselect';
import {
  getFilterGenreSelector,
  getFilterGenreByPropsSelector
} from '../filters/genre';
import {GenreTypes} from '../../constants/genre-type';
import {getPlayedFilmIdSelector} from '../player/player';

export const getFilmsSelector = (state) => state.films.films;
export const getFilmIdSelector = (_, props) => props.match.params.id;
export const getFilmExcludeIdSelector = (_, props) => props && props.excludeId || null;
export const getMaxFilmsCountSelector = (_, props) => props && props.max || null;

export const getFiltratedFilmsSelector = createSelector(
    getFilmsSelector,
    getFilterGenreSelector,
    getFilterGenreByPropsSelector,
    getFilmExcludeIdSelector,
    getMaxFilmsCountSelector,
    (films, genreByState, genreByProps, excludeId, max) => {
      const genreFilter = genreByProps || genreByState;

      if (genreFilter === GenreTypes.ALL) {
        return films;
      }

      const result = films.filter(({genre, id}) => (
        genre === genreFilter && excludeId !== id
      ));

      return max ? result.slice(0, max) : result;
    }
);

export const getFilmByIdSelector = createSelector(
    getFilmsSelector, getFilmIdSelector,
    (films, id) => {
      const parseId = parseInt(id, 10);
      if (isNaN(parseId)) {
        return null;
      }

      return films.find((film) => film.id === parseId);
    }
);

export const getPlayedFilmSelector = createSelector(
    getFilmsSelector, getPlayedFilmIdSelector,
    (films, id) => {
      const parseId = parseInt(id, 10);
      if (isNaN(parseId)) {
        return null;
      }

      return films.find((film) => film.id === parseId);
    }
);
