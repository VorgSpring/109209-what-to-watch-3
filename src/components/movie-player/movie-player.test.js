import React from 'react';
import {MoviePlayerComponent} from './movie-player.jsx';
import renderer from 'react-test-renderer';
import {films} from '../../mocks/films';

describe(`MoviePLayer`, () => {
  it(`should renders correctly`, () => {
    const tree = renderer.create(
        <MoviePlayerComponent
          film={films[0]}
          progress={0}
          currentTime={0}
          loadVideo={()=>{}}
          playVideo={()=>{}}
          refNode={React.createRef()}
          toggleVideo={()=>{}}
          setFullscreen={()=>{}}
          onClosePlayer={()=>{}}
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
