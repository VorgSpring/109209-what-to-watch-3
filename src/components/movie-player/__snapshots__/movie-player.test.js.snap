// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviePLayer should renders correctly with pause status 1`] = `
<div
  className="player"
>
  <video
    autoPlay={true}
    className="player__video"
    controls={false}
    muted={false}
    poster="img/fantastic-beasts-the-crimes-of-grindelwald.jpg"
  >
    <source
      src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
      type="video/mp4"
    />
    <source
      src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
      type="video/webm"
    />
    Your browser doesn't support HTML5 video tag.
  </video>
  <button
    className="player__exit"
    onClick={[Function]}
    type="button"
  >
    Exit
  </button>
  <div
    className="player__controls"
  >
    <div
      className="player__controls-row"
    >
      <div
        className="player__time"
      >
        <progress
          className="player__progress"
          max="100"
          value={0}
        />
        <div
          className="player__toggler"
          style={
            Object {
              "left": "0%",
            }
          }
        >
          Toggler
        </div>
      </div>
      <div
        className="player__time-value"
      >
        0:00:00
      </div>
    </div>
    <div
      className="player__controls-row"
    >
      <button
        className="player__play"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="19"
          viewBox="0 0 19 19"
          width="19"
        >
          <use
            xlinkHref="#play-s"
          />
        </svg>
        <span>
          Play
        </span>
      </button>
      <div
        className="player__name"
      >
        Fantastic Beasts: The Crimes of Grindelwald
      </div>
      <button
        className="player__full-screen"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="27"
          viewBox="0 0 27 27"
          width="27"
        >
          <use
            xlinkHref="#full-screen"
          />
        </svg>
        <span>
          Full screen
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`MoviePLayer should renders correctly with play status 1`] = `
<div
  className="player"
>
  <video
    autoPlay={true}
    className="player__video"
    controls={false}
    muted={false}
    poster="img/fantastic-beasts-the-crimes-of-grindelwald.jpg"
  >
    <source
      src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
      type="video/mp4"
    />
    <source
      src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
      type="video/webm"
    />
    Your browser doesn't support HTML5 video tag.
  </video>
  <button
    className="player__exit"
    onClick={[Function]}
    type="button"
  >
    Exit
  </button>
  <div
    className="player__controls"
  >
    <div
      className="player__controls-row"
    >
      <div
        className="player__time"
      >
        <progress
          className="player__progress"
          max="100"
          value={0}
        />
        <div
          className="player__toggler"
          style={
            Object {
              "left": "0%",
            }
          }
        >
          Toggler
        </div>
      </div>
      <div
        className="player__time-value"
      >
        0:00:00
      </div>
    </div>
    <div
      className="player__controls-row"
    >
      <button
        className="player__play"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="19"
          viewBox="0 0 19 19"
          width="19"
        >
          <use
            xlinkHref="#pause"
          />
        </svg>
        <span>
          Pause
        </span>
      </button>
      <div
        className="player__name"
      >
        Fantastic Beasts: The Crimes of Grindelwald
      </div>
      <button
        className="player__full-screen"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="27"
          viewBox="0 0 27 27"
          width="27"
        >
          <use
            xlinkHref="#full-screen"
          />
        </svg>
        <span>
          Full screen
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`MoviePLayer should renders correctly without status 1`] = `
<div
  className="player"
>
  <video
    autoPlay={true}
    className="player__video"
    controls={false}
    muted={false}
    poster="img/fantastic-beasts-the-crimes-of-grindelwald.jpg"
  >
    <source
      src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
      type="video/mp4"
    />
    <source
      src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
      type="video/webm"
    />
    Your browser doesn't support HTML5 video tag.
  </video>
  <button
    className="player__exit"
    onClick={[Function]}
    type="button"
  >
    Exit
  </button>
</div>
`;
